.github/workflows/test-pr.yml

```diff
@@ -0,0 +1,258 @@
+              exit 1
+            fi
+          fi
+
+  test-k:
+    needs: nix-maven
+    name: 'Build and Test K'
+    runs-on: [self-hosted, linux]
+    steps:
+      - name: 'Check out code'
+        uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: 'Check out k-exercises'
+        uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: 'Set up Docker'
+        run: |
+          docker build . --file Dockerfile.gh --tag runtimeverification/k-ci
+          docker run                        \
+            --name k-ci                     \
+            --rm -it                        \
+            --detach                        \
+            --workdir /opt/workspace        \
+            --user $(id -u):$(id -g)        \
+            -v ${HOME}:${HOME}              \
+            -v "$(pwd):/opt/workspace"      \
+            -v "/etc/passwd:/etc/passwd:ro" \
+            -v "/etc/group:/etc/group:ro"   \
+            runtimeverification/k-ci
+      - name: 'Build K'
+        run: docker exec -t k-ci /bin/bash -c 'mvn package -DskipTests'
+      - name: 'Unit Tests'
+        run: docker exec -t k-ci /bin/bash -c 'mvn --batch-mode verify -U'
+      - name: 'Tutorial Integration Tests'
+        run: |
+          docker exec -t k-ci /bin/bash -c 'k-distribution/target/release/k/bin/spawn-kserver kserver.log'
+          docker exec -t k-ci /bin/bash -c 'cd k-exercises/tutorial && make -j`nproc` --output-sync'
+          docker exec -t k-ci /bin/bash -c 'cd k-distribution/k-tutorial/1_basic && ./test_kompile.sh'
+          docker exec -t k-ci /bin/bash -c 'cd k-distribution/k-tutorial/1_basic && ./test_kompile.sh'
+      - name: 'Tear down Docker'
+        if: always()
+        run: |
+          docker exec -t k-ci /bin/bash -c 'k-distribution/target/release/k/bin/stop-kserver || true'
+          docker stop k-ci
+
+  build-nix-flake:
+    needs: nix-flake-submodules
+    name: 'Build Nix Flake'
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-11]
+    runs-on: ${{ matrix.os }}
+    steps:
+      - name: 'Check out code'
+        uses: actions/checkout@v2.3.4
+        with:
+          # Check out pull request HEAD instead of merge commit.
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: 'Install Nix'
+        uses: cachix/install-nix-action@v15
+        with:
+          extra_nix_config: |
+            substituters = http://cache.nixos.org https://cache.iog.io
+            trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=
+
+      - name: 'Install Cachix'
+        uses: cachix/cachix-action@v10
+        with:
+          name: runtimeverification
+          extraPullNames: kore
+          signingKey: '${{ secrets.CACHIX_SIGNING_KEY }}'
+
+      - name: 'Build K Framework'
+        run: GC_DONT_GC=1 nix build .
+
+  test-nix:
```

Comment:
We could probably try to phase this out in another PR. Most of the steps in `test-nix` (should) have an analogue in `build-nix-flake`

Reply:
Yes, I'd like to optimize the workflows to reduce overall CI time. There seems to be a lot of redundancy in the Nix steps, we can deal with that in another PR though.